include/bluetooth/audio/tbs.h

```diff
@@ -328,4 +328,411 @@ void bt_tbs_register_cb(struct bt_tbs_cb *cbs);
+ *
+ * @param conn          The connection used in the function.
+ * @param err           Error value. BT_TBS_CLIENT_RESULT_CODE_*,
+ *                      GATT error or errno value.
+ * @param inst_index    The index of the TBS instance that was updated.
+ * @param call_count    Number of calls read.
+ * @param calls         Array of calls. The array is not kept by
+ *                      the client, so must be copied to be saved.
+ */
+typedef void (*bt_tbs_client_current_calls_cb)(struct bt_conn *conn, int err,
+					       uint8_t inst_index,
+					       uint8_t call_count,
+					       const struct bt_tbs_client_call *calls);
+
+/**
+ * @brief Callback function for ccp_read_call_state.
+ *
+ * @param conn          The connection used in the function.
+ * @param err           Error value. BT_TBS_CLIENT_RESULT_CODE_*,
+ *                      GATT error or errno value.
+ * @param inst_index    The index of the TBS instance that was updated.
+ * @param call_count    Number of call states read.
+ * @param call_states   Array of call states. The array is not kept by
+ *                      the client, so must be copied to be saved.
+ */
+typedef void (*bt_tbs_client_call_states_cb)(struct bt_conn *conn, int err,
+					     uint8_t inst_index,
+					     uint8_t call_count,
+					     const struct bt_tbs_client_call_state *call_states);
+
+struct bt_tbs_client_cb {
+	bt_tbs_client_discover_cb            discover;
+	bt_tbs_client_cp_cb                  originate_call;
+	bt_tbs_client_cp_cb                  terminate_call;
+	bt_tbs_client_cp_cb                  hold_call;
+	bt_tbs_client_cp_cb                  accept_call;
+	bt_tbs_client_cp_cb                  retrieve_call;
+	bt_tbs_client_cp_cb                  join_calls;
+
+	bt_tbs_client_read_string_cb         bearer_provider_name;
+	bt_tbs_client_read_string_cb         bearer_uci;
+	bt_tbs_client_read_value_cb          technology;
+	bt_tbs_client_read_string_cb         uri_list;
+	bt_tbs_client_read_value_cb          signal_strength;
+	bt_tbs_client_read_value_cb          signal_interval;
+	bt_tbs_client_current_calls_cb       current_calls;
+	bt_tbs_client_read_value_cb          ccid;
+	bt_tbs_client_read_value_cb          status_flags;
+	bt_tbs_client_read_string_cb         call_uri;
+	bt_tbs_client_call_states_cb         call_state;
+	bt_tbs_client_read_value_cb          optional_opcodes;
+	bt_tbs_client_termination_reason_cb  termination_reason;
+	bt_tbs_client_read_string_cb         remote_uri;
+	bt_tbs_client_read_string_cb         friendly_name;
+};
+
+/**
+ * @brief Discover TBS for a connection. This will start a GATT
+ * discover and setup handles and subscriptions.
+ *
+ * @param conn          The connection to discover TBS for.
+ * @param subscribe     Whether to subscribe to all handles.
+ *
+ * @return int          0 on success, GATT error value on fail.
+ */
+int bt_tbs_client_discover(struct bt_conn *conn, bool subscribe);
```

Comment:
No such thing as a stupid question ^^ Why not subscribe by default? Most of the TBS characteristics have Notify property, thus once read *seamlessly* by client implementation right after connection could be notified if there's any change by the server.

This would reduce this API by removing all the read request functions like bt_tbs_client_read_bearer_provider_name.
If there's possiblity to notify I would stick to it as ATT Read operation is blocking while ATT Notification can be sent anytime.

Reply:
This is some very old code where we wanted to give that flexibility. It is likely something we will remove, and always subscribe by default. Since this is just moving the current implementation, this won't be fixed in this PR, but will be fixed in a future update.

Reply:
https://github.com/zephyrproject-rtos/zephyr/issues/42488